photon-core/src/main/java/org/photonvision/vision/pipe/impl/CropPipe.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) Photon Vision.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed mat the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package org.photonvision.vision.pipe.impl;
+
+import edu.wpi.first.math.MathUtil;
+import org.opencv.core.Mat;
+import org.opencv.core.Point;
+import org.opencv.core.Rect;
+import org.photonvision.vision.opencv.CVMat;
+import org.photonvision.vision.pipe.CVPipe;
+
+public class CropPipe extends CVPipe<CVMat, CVMat, Rect> {
+    public Rect dynamiRect;
+
+    public CropPipe() {
+        Rect full_screen = new Rect(0, 0, Integer.MAX_VALUE, Integer.MAX_VALUE);
+
+        this.params = full_screen;
+        dynamiRect = full_screen;
+    }
+
+    @Override
+    protected CVMat process(CVMat in) {
+        Mat mat = in.getMat();
+        // if (fullyCovers(params, mat)) {
+        //     return in;
+        // }
+        Rect proccesor = false ? this.params : dynamiRect;
+
+        int x = MathUtil.clamp(proccesor.x, 0, mat.width());
+        int y = MathUtil.clamp(proccesor.y, 0, mat.height());
+        int width = MathUtil.clamp(proccesor.width, 0, mat.width() - x);
+        int height = MathUtil.clamp(proccesor.height, 0, mat.height() - y);
+
+        return new CVMat(mat.submat(y, y + height, x, x + width));
```

Comment:
The submat here is just a submat of the overall data array, and does not actually own the native memory. but it still needs to be released. So we need to make sure that we `release` both after we are done with them. I'd argue we should release the cropped Mat immediately after Apriltag detection, since we don't use it anywhere else. See this example:


```java
   @Test
    public void testRelease() {
        var pipe = new CropPipe();
        pipe.setParams(new Rect(10, 10, 50, 50));

        var frameProvider =
                new FileFrameProvider(
                        TestUtils.getApriltagImagePath(TestUtils.ApriltagTestImages.kTag1_640_480, false),
                        TestUtils.WPI2020Image.FOV,
                        TestUtils.get2020LifeCamCoeffs(false));
        frameProvider.requestFrameThresholdType(FrameThresholdType.GREYSCALE);

        var frame = frameProvider.get();
        var out = pipe.run(frame.processedImage);

        System.out.println(frame.processedImage.getMat());
        System.out.println(out.output.getMat());

        Imgcodecs.imwrite("pre_release_full.png", frame.processedImage.getMat());
        Imgcodecs.imwrite("pre_release_crop.png", out.output.getMat());
        
        frame.release();

        System.out.println(frame.processedImage.getMat());
        System.out.println(out.output.getMat());

        Imgcodecs.imwrite("pose_releaseframe_crop.png", out.output.getMat());

        out.output.getMat().release();

        System.out.println(frame.processedImage.getMat());
        System.out.println(out.output.getMat());
    }
```

Output is:
```
// Starting state
    Mat [ 480*640*CV_8UC1, isCont=true, isSubmat=false, nativeObj=0x7ff994f80a30, dataAddr=0x7ff994fb26c0 ]
    Mat [ 50*50*CV_8UC1, isCont=false, isSubmat=true, nativeObj=0x7ff994f87910, dataAddr=0x7ff994fb3fca ]
// After releasing the full image Mat
    Mat [ 0*0*CV_8UC1, isCont=true, isSubmat=false, nativeObj=0x7ff994f80a30, dataAddr=0x0 ]
    Mat [ 50*50*CV_8UC1, isCont=false, isSubmat=true, nativeObj=0x7ff994f87910, dataAddr=0x7ff994fb3fca ]
// after releasing both
    Mat [ 0*0*CV_8UC1, isCont=true, isSubmat=false, nativeObj=0x7ff994f80a30, dataAddr=0x0 ]
    Mat [ 0*0*CV_8UC1, isCont=false, isSubmat=true, nativeObj=0x7ff994f87910, dataAddr=0x0 ]
```